src/ColorPicker/Halogen/Utils/Drag.purs

```diff
@@ -0,0 +1,206 @@
+module ColorPicker.Halogen.Utils.Drag
+  ( dragEventSource
+  , DragData
+  , CursorEvent
+  , DragEvent(..)
+  , DragEffects
+  , Position
+  , cursorEventToPosition
+  , cursorEventToTarget
+  , mkFirstDragData
+  , mkDragData
+  ) where
+
+import Prelude
+
+import Control.Alt ((<|>))
+import Control.Monad.Aff.AVar (AVAR)
+import Control.Monad.Aff.Class (class MonadAff)
+import Control.Monad.Eff (Eff)
+import Control.Monad.Eff.Ref (REF, newRef, readRef, writeRef)
+import Control.Monad.Except (runExcept)
+import DOM (DOM)
+import DOM.Classy.Event (target)
+import DOM.Classy.HTMLElement (getBoundingClientRect, DOMRect)
+import DOM.Classy.Node (fromNode)
+import DOM.Event.EventTarget (EventListener, eventListener, addEventListener, removeEventListener)
+import DOM.Event.MouseEvent as MouseE
+import DOM.Event.TouchEvent as TouchE
+import DOM.Event.Types (Event, EventType)
+import DOM.HTML (window)
+import DOM.HTML.Event.EventTypes (mousemove, mouseup, touchend, touchmove)
+import DOM.HTML.Types (HTMLElement, windowToEventTarget)
+import DOM.HTML.Window (scrollX, scrollY)
+import DOM.Node.Types (Node)
+import Data.Either (Either(..), either, hush)
+import Data.Foldable (traverse_)
+import Data.Int (toNumber)
+import Data.Maybe (Maybe(..), fromMaybe)
+import Halogen.Query.EventSource as ES
+
+
+type DragData =
+  { page ∷ Position
+  , progress ∷ Position
+  , pointer ∷ Position
+  , delta ∷ Position
+  , offset ∷ Position
+  }
+
+type CursorEvent = Either MouseE.MouseEvent TouchE.TouchEvent
```

Comment:
Defining `CursorEvent` using `MouseE.MouseEvent` and `TouchE.TouchEvent` describes to client what type of event we support we it can only supports this two types. and client might be interested to know that CursorEvent is MouseEvent or TouchEvent.